File: app/services/plaid_service.py
import plaid
from plaid.api import plaid_api
from plaid.model.transactions_get_request import TransactionsGetRequest
from plaid.model.accounts_get_request import AccountsGetRequest
from plaid.model.item_public_token_exchange_request import ItemPublicTokenExchangeRequest
from plaid.model.link_token_create_request import LinkTokenCreateRequest
from plaid.model.link_token_create_request_user import LinkTokenCreateRequestUser
from plaid.model.country_code import CountryCode
from plaid.model.products import Products
from plaid.model.institutions_get_by_id_request import InstitutionsGetByIdRequest
from plaid.model.transactions_sync_request import TransactionsSyncRequest
from plaid.configuration import Configuration
from plaid.api_client import ApiClient
from datetime import datetime, timedelta
from typing import Optional, List, Dict, Any
import logging

from app.core.config import settings

logger = logging.getLogger(__name__)


class PlaidService:
    def __init__(self):
        # Configure Plaid client
        configuration = Configuration(
            host=self._get_plaid_host(),
            api_key={
                'clientId': settings.plaid_client_id,
                'secret': settings.plaid_secret,
            }
        )
        api_client = ApiClient(configuration)
        self.client = plaid_api.PlaidApi(api_client)

    def _get_plaid_host(self):
        """Get Plaid host based on environment"""
        if settings.plaid_env == "sandbox":
            return plaid.Environment.sandbox
        elif settings.plaid_env == "development":
            return plaid.Environment.development
        else:
            return plaid.Environment.production

    async def create_link_token(self, user_id: int, webhook_url: Optional[str] = None) -> Dict[str, Any]:
        """Create a link token for Plaid Link"""
        try:
            request = LinkTokenCreateRequest(
                products=[Products('transactions'), Products('auth')],
                client_name="Financial Transaction Aggregator",
                country_codes=[CountryCode('US')],
                language='en',
                webhook=webhook_url,
                user=LinkTokenCreateRequestUser(
                    client_user_id=str(user_id)
                )
            )
            
            response = self.client.link_token_create(request)
            return {
                "link_token": response['link_token'],
                "expiration": response['expiration'].isoformat(),
                "request_id": response['request_id']
            }
        except Exception as e:
            logger.error(f"Error creating link token: {str(e)}")
            raise

    async def exchange_public_token(self, public_token: str) -> Dict[str, str]:
        """Exchange public token for access token"""
        try:
            request = ItemPublicTokenExchangeRequest(
                public_token=public_token
            )
            
            response = self.client.item_public_token_exchange(request)
            return {
                "access_token": response['access_token'],
                "item_id": response['item_id']
            }
        except Exception as e:
            logger.error(f"Error exchanging public token: {str(e)}")
            raise

    async def get_accounts(self, access_token: str) -> List[Dict[str, Any]]:
        """Get accounts for an access token"""
        try:
            request = AccountsGetRequest(access_token=access_token)
            response = self.client.accounts_get(request)
            
            accounts = []
            for account in response['accounts']:
                accounts.append({
                    "account_id": account['account_id'],
                    "name": account['name'],
                    "official_name": account.get('official_name'),
                    "type": account['type'],
                    "subtype": account['subtype'],
                    "mask": account.get('mask'),
                    "balances": {
                        "available": account['balances'].get('available'),
                        "current": account['balances'].get('current'),
                        "limit": account['balances'].get('limit'),
                        "iso_currency_code": account['balances'].get('iso_currency_code'),
                        "unofficial_currency_code": account['balances'].get('unofficial_currency_code')
                    },
                    "verification_status": account.get('verification_status'),
                    "persistent_account_id": account.get('persistent_account_id')
                })
            
            return accounts, response['item']
        except Exception as e:
            logger.error(f"Error getting accounts: {str(e)}")
            raise

    async def get_transactions(
        self, 
        access_token: str, 
        start_date: datetime, 
        end_date: datetime,
        account_ids: Optional[List[str]] = None,
        count: int = 500,
        offset: int = 0
    ) -> List[Dict[str, Any]]:
        """Get transactions for an access token"""
        try:
            request = TransactionsGetRequest(
                access_token=access_token,
                start_date=start_date.date(),
                end_date=end_date.date(),
                account_ids=account_ids,
                count=count,
                offset=offset
            )
            
            response = self.client.transactions_get(request)
            
            transactions = []
            for transaction in response['transactions']:
                transactions.append({
                    "transaction_id": transaction['transaction_id'],
                    "account_id": transaction['account_id'],
                    "amount": transaction['amount'],
                    "iso_currency_code": transaction.get('iso_currency_code'),
                    "unofficial_currency_code": transaction.get('unofficial_currency_code'),
                    "category": transaction.get('category', []),
                    "category_id": transaction.get('category_id'),
                    "merchant_name": transaction.get('merchant_name'),
                    "name": transaction['name'],
                    "original_description": transaction.get('original_description'),
                    "date": transaction['date'].isoformat(),
                    "authorized_date": transaction.get('authorized_date').isoformat() if transaction.get('authorized_date') else None,
                    "authorized_datetime": transaction.get('authorized_datetime').isoformat() if transaction.get('authorized_datetime') else None,
                    "datetime": transaction.get('datetime').isoformat() if transaction.get('datetime') else None,
                    "pending": transaction.get('pending', False),
                    "pending_transaction_id": transaction.get('pending_transaction_id'),
                    "account_owner": transaction.get('account_owner'),
                    "transaction_type": transaction.get('transaction_type'),
                    "transaction_code": transaction.get('transaction_code'),
                    "location": transaction.get('location', {}),
                    "payment_meta": transaction.get('payment_meta', {}),
                    "personal_finance_category": transaction.get('personal_finance_category', {})
                })
            
            return transactions, response['total_transactions']
        except Exception as e:
            logger.error(f"Error getting transactions: {str(e)}")
            raise

    async def sync_transactions(self, access_token: str, cursor: Optional[str] = None) -> Dict[str, Any]:
        """Sync transactions using cursor-based pagination"""
        try:
            request = TransactionsSyncRequest(
                access_token=access_token,
                cursor=cursor
            )
            
            response = self.client.transactions_sync(request)
            
            # Process added transactions
            added_transactions = []
            for transaction in response['added']:
                added_transactions.append({
                    "transaction_id": transaction['transaction_id'],
                    "account_id": transaction['account_id'],
                    "amount": transaction['amount'],
                    "iso_currency_code": transaction.get('iso_currency_code'),
                    "unofficial_currency_code": transaction.get('unofficial_currency_code'),
                    "category": transaction.get('category', []),
                    "category_id": transaction.get('category_id'),
                    "merchant_name": transaction.get('merchant_name'),
                    "name": transaction['name'],
                    "original_description": transaction.get('original_description'),
                    "date": transaction['date'].isoformat(),
                    "authorized_date": transaction.get('authorized_date').isoformat() if transaction.get('authorized_date') else None,
                    "authorized_datetime": transaction.get('authorized_datetime').isoformat() if transaction.get('authorized_datetime') else None,
                    "datetime": transaction.get('datetime').isoformat() if transaction.get('datetime') else None,
                    "pending": transaction.get('pending', False),
                    "pending_transaction_id": transaction.get('pending_transaction_id'),
                    "account_owner": transaction.get('account_owner'),
                    "transaction_type": transaction.get('transaction_type'),
                    "transaction_code": transaction.get('transaction_code'),
                    "location": transaction.get('location', {}),
                    "payment_meta": transaction.get('payment_meta', {}),
                    "personal_finance_category": transaction.get('personal_finance_category', {})
                })
            
            # Process modified transactions
            modified_transactions = []
            for transaction in response['modified']:
                modified_transactions.append({
                    "transaction_id": transaction['transaction_id'],
                    "account_id": transaction['account_id'],
                    "amount": transaction['amount'],
                    # ... same fields as added transactions
                })
            
            # Process removed transaction IDs
            removed_transaction_ids = [t['transaction_id'] for t in response['removed']]
            
            return {
                "added": added_transactions,
                "modified": modified_transactions,
                "removed": removed_transaction_ids,
                "next_cursor": response['next_cursor'],
                "has_more": response['has_more']
            }
        except Exception as e:
            logger.error(f"Error syncing transactions: {str(e)}")
            raise

    async def get_institution(self, institution_id: str) -> Dict[str, Any]:
        """Get institution details by ID"""
        try:
            request = InstitutionsGetByIdRequest(
                institution_id=institution_id,
                country_codes=[CountryCode('US')]
            )
            
            response = self.client.institutions_get_by_id(request)
            institution = response['institution']
            
            return {
                "institution_id": institution['institution_id'],
                "name": institution['name'],
                "country": institution.get('country'),
                "url": institution.get('url'),
                "primary_color": institution.get('primary_color'),
                "logo": institution.get('logo')
            }
        except Exception as e:
            logger.error(f"Error getting institution: {str(e)}")
            raise

File: financial-ai-app/backend/services/plaidService.js
const { Configuration, PlaidApi, PlaidEnvironments } = require('plaid');
require('dotenv').config();

// Initialize Plaid client
const configuration = new Configuration({
  basePath: PlaidEnvironments[process.env.PLAID_ENV || 'sandbox'],
  baseOptions: {
    headers: {
      'PLAID-CLIENT-ID': process.env.PLAID_CLIENT_ID,
      'PLAID-SECRET': process.env.PLAID_SECRET,
    },
  },
});

const plaidClient = new PlaidApi(configuration);

class PlaidService {
  // Create a Link token for connecting bank accounts
  async createLinkToken(userId) {
    try {
      const request = {
        user: {
          client_user_id: userId,
        },
        client_name: 'Financial AI App',
        products: ['transactions', 'accounts', 'liabilities'],
        country_codes: ['US'],
        language: 'en',
        webhook: `${process.env.BACKEND_URL || 'http://localhost:5000'}/api/webhooks/plaid`,
      };

      const response = await plaidClient.linkTokenCreate(request);
      return response.data;
    } catch (error) {
      console.error('Error creating link token:', error);
      throw new Error('Failed to create link token');
    }
  }

  // Exchange public token for access token
  async exchangePublicToken(publicToken) {
    try {
      const request = {
        public_token: publicToken,
      };

      const response = await plaidClient.itemPublicTokenExchange(request);
      return response.data;
    } catch (error) {
      console.error('Error exchanging public token:', error);
      throw new Error('Failed to exchange public token');
    }
  }

  // Get accounts for an item
  async getAccounts(accessToken) {
    try {
      const request = {
        access_token: accessToken,
      };

      const response = await plaidClient.accountsGet(request);
      return response.data;
    } catch (error) {
      console.error('Error fetching accounts:', error);
      throw new Error('Failed to fetch accounts');
    }
  }

  // Get transactions for an item
  async getTransactions(accessToken, startDate, endDate, accountIds = null) {
    try {
      const request = {
        access_token: accessToken,
        start_date: startDate,
        end_date: endDate,
        options: {
          include_personal_finance_category: true,
          include_logo_and_counterparty_beta: true,
        },
      };

      if (accountIds) {
        request.account_ids = accountIds;
      }

      const response = await plaidClient.transactionsGet(request);
      return response.data;
    } catch (error) {
      console.error('Error fetching transactions:', error);
      throw new Error('Failed to fetch transactions');
    }
  }

  // Get account balances
  async getBalances(accessToken) {
    try {
      const request = {
        access_token: accessToken,
      };

      const response = await plaidClient.accountsBalanceGet(request);
      return response.data;
    } catch (error) {
      console.error('Error fetching balances:', error);
      throw new Error('Failed to fetch balances');
    }
  }

  // Get item (connection) information
  async getItem(accessToken) {
    try {
      const request = {
        access_token: accessToken,
      };

      const response = await plaidClient.itemGet(request);
      return response.data;
    } catch (error) {
      console.error('Error fetching item info:', error);
      throw new Error('Failed to fetch item information');
    }
  }

  // Update webhook for an item
  async updateWebhook(accessToken, webhookUrl) {
    try {
      const request = {
        access_token: accessToken,
        webhook: webhookUrl,
      };

      const response = await plaidClient.itemWebhookUpdate(request);
      return response.data;
    } catch (error) {
      console.error('Error updating webhook:', error);
      throw new Error('Failed to update webhook');
    }
  }

  // Remove an item (disconnect account)
  async removeItem(accessToken) {
    try {
      const request = {
        access_token: accessToken,
      };

      const response = await plaidClient.itemRemove(request);
      return response.data;
    } catch (error) {
      console.error('Error removing item:', error);
      throw new Error('Failed to remove item');
    }
  }

  // Get institutions by name
  async searchInstitutions(query, products = ['transactions'], countryCodes = ['US']) {
    try {
      const request = {
        query,
        products,
        country_codes: countryCodes,
      };

      const response = await plaidClient.institutionsSearch(request);
      return response.data;
    } catch (error) {
      console.error('Error searching institutions:', error);
      throw new Error('Failed to search institutions');
    }
  }

  // Get institution by ID
  async getInstitution(institutionId) {
    try {
      const request = {
        institution_id: institutionId,
        country_codes: ['US'],
      };

      const response = await plaidClient.institutionsGetById(request);
      return response.data;
    } catch (error) {
      console.error('Error fetching institution:', error);
      throw new Error('Failed to fetch institution');
    }
  }

  // Sync transactions (for periodic updates)
  async syncTransactions(accessToken, cursor = null) {
    try {
      const request = {
        access_token: accessToken,
      };

      if (cursor) {
        request.cursor = cursor;
      } else {
        request.options = {
          include_personal_finance_category: true,
          include_logo_and_counterparty_beta: true,
        };
      }

      const response = await plaidClient.transactionsSync(request);
      return response.data;
    } catch (error) {
      console.error('Error syncing transactions:', error);
      throw new Error('Failed to sync transactions');
    }
  }
}

module.exports = new PlaidService();